jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+//| Requires
+//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+== Abstract
+
+link:https://github.com/jenkinsci/jep/tree/master/jep/300:[Jenkins Essentials], and more specifically its link:https://github.com/jenkinsci/jep/tree/master/jep/301:[Evergreen] _component_, aims at providing an link:https://github.com/jenkinsci/jep/tree/master/jep/300#auto-update[automatically updating distribution] of Jenkins.
+
+Continuous Delivery is about making small incremental changes, making failures much more easily recoverable. In the context here, it means we need to be able to seamlessly upgrade Jenkins, _but also_ roll back to the previously running version if an upgrade goes wrong.
+As Jenkins does not support downgrading alone, we need to introduce a system that will enable that auto-downgrade capability.
+
+This document outlines the design of the snapshotting system we plan to use.
+
+== Specification
+
+_Evergreen_ works with two main components: Jenkins itself, and what we call the evergreen-client.
+
+=== Upgrading/downgrading
+
+Once an evergreen client has been instructed to perform an upgrade, it is responsible for the following operations:
+
+1. (If needed) Initialize the git repository.
+2. Stop Jenkins
+3. Take a snapshot (see <<snapshot>> below)
+5. Perform the instructed upgrade to the given Essentials BOM
+footnote:[Bill Of Materials: this format is currently being designed, but will list everything constituting a version of Essentials: WAR and exact versions of all plugins]
+6. Start new Jenkins version
+7. [[healthcheck]] Check it is responding in a healthy manner.
+8. If not:
+.. Take a snapshot footnote:[this way, if new files were created, we don't just delete them in an unrecoverable way when going back to the previous snapshot].
+.. Rollback to the previous data snapshot and Essentials BOM version.
+To do so, we will generally create a new commit (i.e. avoid `git reset --hard HEAD~`), to keep a durable track of where we went through, accessible through `git log`.
+.. Start (previous) Jenkins version
+// what if starting the previous version doesn't work either?
+9. Report the outcome to the Essentials backend.
+
+// Need to create some diagram to help quickly grok/understand the long list above
+
+[[snapshot]]
+==== Take a snapshot
+
+We use `git` for this purpose, so in this context:
+
+////
+I think .gitignore content must be designed to be able to evolve over time.
+To allow more flexibility, I think the content should be associated between an essentials release/bom to a given .gitignore content.
+////
+
+* Update `.gitignore` content with current Essentials release.
+* `git add .`
+* `git add -u `
```

Comment:
Yes, I mixed things up here. I was pretty sure `git add .` didn't handle deleted files, so that's why I did two lines. 

In any case, fixed to use `git add --all` which also makes the intent clearer that we want _all_ files, added, modified, or removed.